added new strain rate and derivative functions to diffusion and dislocation creep

bobmyhill opened the pull request:
This PR adds `strain rate(P, T, C, stress)` and `d(strain rate)/d(stress) (P, T, C, stress)` functions to diffusion and dislocation creep. These stress-dependent functions are needed for correct compositing (via a Newton-Raphson iteration). 

At the moment, diffusion_dislocation.cc contains these functions, but more composite rheologies are now being considered, and each of these will need the same functions.

Opened for comment, especially on the use of std::map<std::string, double> to store the creep parameters. The storage avoids repeated calculation of the creep parameters in compute_strain_rate_and_derivative(), but may slow down the code. If there's a more clean and efficient way to do this, I'll make a change.

bobmyhill commented:
@tjhei, previous discussions suggest that you might want to comment on this kind of thing :)

gassmoeller commented:
Oh, and if your main goal is to get the derivative function in (which I like and want), then starting with a slow, but uncontroversial version is easier. But this needs a test case where the derivative is computed/used in some form.

bobmyhill commented:
Hi @gassmoeller,
What do you think about this? I've added a params struct for diffusion and dislocation creep, and changed the diffusion_dislocation material model so that it is a composite of those two creep mechanisms.

gassmoeller commented on include/aspect/material_model/rheology/diffusion_creep.h:
```diff
@@ -62,6 +78,19 @@ namespace aspect
+           * Compute the creep parameters for the diffusion creep law.
+           * If @p expected_n_phases_per_composition points to a vector of
+           * unsigned integers this is considered the number of phase transitions
+           * for each compositional field and viscosity will be fisrt computed on
```
```suggestion
           * for each compositional field and viscosity will be first computed on
```

bobmyhill replied:
Ooops!

gassmoeller commented on include/aspect/material_model/rheology/diffusion_creep.h:
```diff
@@ -62,6 +78,19 @@ namespace aspect
+           * If @p expected_n_phases_per_composition points to a vector of
+           * unsigned integers this is considered the number of phase transitions
+           * for each compositional field and viscosity will be fisrt computed on
+           * each phases and then averaged for each compositional field.
```
```suggestion
           * each phase and then averaged for each compositional field.
```

bobmyhill replied:
Done.

gassmoeller commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -39,52 +39,91 @@ namespace aspect
         return viscosity_diffusion;
       }
 
 
```
add another empty line (3 lines between functions), same below this function

bobmyhill replied:
Done (also in other places here and in dislocation_creep.cc

gassmoeller commented on source/material_model/rheology/dislocation_creep.cc:
```diff
@@ -37,54 +37,93 @@ namespace aspect
       {}
 
 
-
```
Keep empty line

bobmyhill replied:
Done.

gassmoeller commented on source/material_model/rheology/dislocation_creep.cc:
```diff
@@ -37,54 +37,93 @@ namespace aspect
+
+        return creep_parameters;
+      }
+
```
3 empty lines, same below the function

bobmyhill replied:
Done.

gassmoeller commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -39,52 +39,91 @@ namespace aspect
+
+        const double strain_rate_diffusion = stress * dstrain_rate_dstress_diffusion;
+
+        const std::pair<double, double> strain_rate_and_derivative (strain_rate_diffusion, dstrain_rate_dstress_diffusion);
+        return strain_rate_and_derivative;
```
this can be somewhat simplified in a rare case of shorter *and* simpler to understand :-)

```suggestion
        return std::make_pair(strain_rate_diffusion, dstrain_rate_dstress_diffusion);
```

bobmyhill replied:
aha!

gassmoeller commented on source/material_model/rheology/dislocation_creep.cc:
```diff
@@ -37,54 +37,93 @@ namespace aspect
+                                                        std::exp(-(creep_parameters.activation_energy + pressure*creep_parameters.activation_volume)/
+                                                                 (constants::gas_constant*temperature));
+
+        const std::pair<double, double> strain_rate_and_derivative (strain_rate_dislocation, dstrain_rate_dstress_dislocation);
+        return strain_rate_and_derivative;
```
same simplification as in diffusion_creep

bobmyhill replied:
Yup. Done

gassmoeller commented:
/rebuild
